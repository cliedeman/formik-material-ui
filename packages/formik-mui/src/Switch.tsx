import * as React from 'react';
import {
  type SwitchProps as MuiSwitchProps,
  Switch as MuiSwitch, 
} from '@mui/material';
import type { FieldProps } from 'formik';
import invariant from 'tiny-warning';

export interface SwitchProps
  extends FieldProps,
    Omit<
      MuiSwitchProps,
      | 'checked'
      | 'name'
      | 'value'
      | 'defaultChecked'
      | 'form'
      // Excluded for conflict with Field type
      | 'type'
    > {
  type?: string;
}

export function fieldToSwitch({
  disabled,
  field: { onBlur: fieldOnBlur, ...field },
  form: { isSubmitting },
  type,
  onBlur,
  ...props
}: SwitchProps): MuiSwitchProps {
  if (process.env.NODE_ENV !== 'production') {
    invariant(
      type === 'checkbox',
      `property type=checkbox is missing from field ${field.name}, this can caused unexpected behaviour`
    );
  }

  return {
    disabled: disabled ?? isSubmitting,
    onBlur:
      onBlur ??
      function (e) {
        fieldOnBlur(e ?? field.name);
      },
    ...field,
    ...props,
  };
}

export function Switch(props: SwitchProps) {
  return <MuiSwitch {...fieldToSwitch(props)} />;
}

Switch.displayName = 'FormikMaterialUISwitch';
