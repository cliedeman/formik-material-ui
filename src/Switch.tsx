// @ts-ignore
import * as React from 'react';
import MuiSwitch, {
  SwitchProps as MuiSwitchProps,
} from '@material-ui/core/Switch';
import { FieldProps } from 'formik';
import { Omit } from './types';

export interface SwitchProps extends FieldProps, Omit<MuiSwitchProps, 'form'> {}

const Switch: React.ComponentType<SwitchProps> = ({
  field,
  form: { isSubmitting },
  disabled = false,
  ...props
}) => (
  <MuiSwitch
    disabled={isSubmitting || disabled}
    {...props}
    {...field}
    // TODO: is this the correct way?
    value={field.value ? 'checked' : ''}
  />
);

Switch.displayName = 'FormikMaterialUISwitch';

export default Switch;
