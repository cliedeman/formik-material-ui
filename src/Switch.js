/* @flow */
import Switch from '@material-ui/core/Switch';

import createComponent from './createComponent';

export default createComponent(
  Switch,
  ({ field, form: { isSubmitting }, disabled = false, ...props }) => ({
    disabled: isSubmitting || disabled,
    ...props,
    ...field,
    // TODO: is this the correct way?
    value: field.value ? 'checked' : '',
  })
);
