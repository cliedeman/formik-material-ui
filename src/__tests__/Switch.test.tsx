import * as React from 'react';
import { Formik, Field, Form } from 'formik';
import renderer from 'react-test-renderer';

import { Switch } from '../Switch';

test('Switch Renders Correctly', () => {
  const component = renderer.create(
    <Formik initialValues={{}} onSubmit={() => {}}>
      <Form>
        <Field name="test" label="Text" component={Switch} />
      </Form>
    </Formik>
  );

  expect(component.toJSON()).toMatchSnapshot();
});
