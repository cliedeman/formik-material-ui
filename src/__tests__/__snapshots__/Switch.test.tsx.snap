// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch Renders Correctly 1`] = `
<form
  onSubmit={[Function]}
>
  <span
    className="MuiSwitch-root-1"
  >
    <span
      className="MuiButtonBase-root-20 MuiIconButton-root-14 MuiSwitchBase-root-10 MuiSwitch-switchBase-4 MuiSwitch-colorSecondary-7"
      label="Text"
      onBlur={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={null}
    >
      <span
        className="MuiIconButton-label-19"
      >
        <span
          className="MuiSwitch-icon-2"
        />
        <input
          checked={false}
          className="MuiSwitchBase-input-13"
          disabled={false}
          name="test"
          onChange={[Function]}
          type="checkbox"
        />
      </span>
      <span
        className="MuiTouchRipple-root-23"
      />
    </span>
    <span
      className="MuiSwitch-bar-9"
    />
  </span>
</form>
`;
